File: frontend/src/api/realtime.ts
import { apiRequest } from "./client";

export type RealtimeSettings = {
  enabled: boolean;
  interval_seconds: number;
  batch_size: number;
  max_concurrency: number;
  min_interval_seconds: number;
  last_run_at: string | null;
  next_run_at: string | null;
  updated_at: string;
};

export type ProtectionJob = {
  id: string;
  trigger: "manual" | "scheduled";
  status: "queued" | "running" | "succeeded" | "partial" | "failed" | "skipped";
  total_services: number;
  processed_services: number;
  succeeded_services: number;
  failed_services: number;
  new_notifications: number;
  current_batch: number;
  current_service_names?: string[];
  retry_count: number;
  error_summary: { code?: string; message?: string; service_id?: string }[];
  started_at: string | null;
  heartbeat_at: string | null;
  completed_at: string | null;
  created_at: string;
};

export function getRealtimeSettings(signal?: AbortSignal): Promise<RealtimeSettings> {
  return apiRequest("/v1/settings/realtime-protection", { signal }, { authenticated: true });
}

export function updateRealtimeSettings(
  input: Partial<Pick<RealtimeSettings, "enabled" | "interval_seconds" | "batch_size" | "max_concurrency">>,
): Promise<RealtimeSettings> {
  return apiRequest("/v1/settings/realtime-protection", { method: "PATCH", body: JSON.stringify(input) }, { authenticated: true });
}

export function runRealtimeProtection(): Promise<ProtectionJob> {
  return apiRequest("/v1/settings/realtime-protection/run-now", { method: "POST" }, { authenticated: true });
}

export function getCurrentProtectionJob(signal?: AbortSignal): Promise<ProtectionJob | null> {
  return apiRequest("/v1/settings/realtime-protection/current-job", { signal }, { authenticated: true });
}

File: backend/app/schemas/realtime.py
from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, Field


class RealtimeSettingsUpdate(BaseModel):
    enabled: bool | None = None
    interval_seconds: int | None = Field(default=None, ge=60, le=31_536_000)
    batch_size: int | None = Field(default=None, ge=1, le=100)
    max_concurrency: int | None = Field(default=None, ge=1, le=10)


class RealtimeSettingsResponse(BaseModel):
    enabled: bool
    interval_seconds: int
    batch_size: int
    max_concurrency: int
    min_interval_seconds: int
    last_run_at: datetime | None
    next_run_at: datetime | None
    updated_at: datetime


class ProtectionJobResponse(BaseModel):
    id: UUID
    trigger: Literal["manual", "scheduled"]
    status: Literal["queued", "running", "succeeded", "partial", "failed", "skipped"]
    total_services: int
    processed_services: int
    succeeded_services: int
    failed_services: int
    new_notifications: int
    current_batch: int
    current_service_names: list[str] = Field(default_factory=list)
    retry_count: int
    error_summary: list
    started_at: datetime | None
    heartbeat_at: datetime | None
    completed_at: datetime | None
    created_at: datetime

    model_config = {"from_attributes": True}
